substrate/client/transaction-pool/src/fork_aware_txpool/mod.rs

```
@@ -378,6 +378,44 @@ mod view_store;
 
 pub use fork_aware_txpool::{ForkAwareTxPool, ForkAwareTxPoolTask};
 
+// Holds various helper types useful in determining transaction validity.
+mod transaction_validation_util {
```

Should be removed. Metrics are generally not for debugging or tracing, just to give an insight to the system.

---

Generally these internal errors should not be exposed in this way. I'm slightly okay with exposing any kind of transaction validity status to the metrics, which is already discussable. Because I could just send transaction that are always invalid and you will get 0 insights from this. However, these internal errors should not be exposed. Maybe the code should be changed that transactions are not invalid or whatever when there was an internal error. Aka retry with a different block or whatever to verify the transaction.

---

An addition to all of the above:  https://github.com/paritytech/polkadot-sdk/pull/8640#discussion_r2111313009 mentions that custom invalid errors are used only for testing. That means we shouldn't propagate them up in the `invalid_txs` set, so we'd need to return a tuple of invalid txs with proper validity errors + internal validation error txs that miss the actual error - we could propagate these errors too, but we'd need to change `TransactionStatus::Invalid` so that it wraps a message that can be lended out to users by the RPC API. Either way I would have this decision on its own, not here. For now my suggestion above still stands.

---

My 2 cents.

> I'm slightly okay with exposing any kind of transaction validity status to the metrics, which is already discussable. 

Yeah, was hesitant about this too. 

The reason why I want to introduce it is because the removal of invalid transcations during _internal mempool revalidation_ (which is triggered only on finalized-block events) should not really happen. Invalid transactions should be removed earlier - either at initial submission validation or during view revalidation or while block building.

This however happens for every prod and test net - and the reason is not clear for me. I haven't observed this on local zombienet testing. I'd like to better understand what is going on, and IMO metrics are better approach then logging for this case.

These metrics are rather for the node/txpool maintainers (read: node team and its successors) rather than for node operators. They should not really be interested in this particular detail, and they have other metrics to look into.


> Because I could just send transaction that are always invalid and you will get 0 insights from this. 

As I explained above - an invalid transaction should be removed earlier. If you know how to construct transaction that will be removed on the mempool-revalidation stage, please share - maybe I am missing something. I tried to reason about this, but did not figure it out.


> However, these internal errors should not be exposed. Maybe the code should be changed that transactions are not invalid or whatever when there was an internal error.

I hope that above justifies why this is also exposed - I need to know the reason why transactions are reported as invalid (and internal error might be one of the reasons). 
I agree - if transactions are dropped due to internal error (uknown), maybe we need to improve implementation. But first I need to know what is happening.

Overall - I am not 100% convinced to exposing metrics (however I think long-term it is good to have them). This could also be debugged by just adding some logs and enabling them on some nodes (but I think using metrics/labels is easier and more reusable).

---

> > I've just realize that my comment was not correct. Any runtime could return this value during transaction validation 🙈.
> 
> Yeah and it is also used.
> 
> Could we still maybe expose these internal errors in a different way than trying to hack them into the `TransactionValidity`?

oooh, I see the problem :)

We have a bit of duplication between internal txpool structure [here](https://github.com/paritytech/polkadot-sdk/blob/436b4935b52562f79a83b6ecadeac7dcbc1c2367/substrate/client/transaction-pool/src/graph/validated_pool.rs#L53-L62) and [this](https://github.com/paritytech/polkadot-sdk/blob/436b4935b52562f79a83b6ecadeac7dcbc1c2367/substrate/primitives/runtime/src/transaction_validity.rs#L156-L161).

Best way to go would be add a new variant (e.g. `ValidationFailed`) to [`ValidatedTransaction`](https://github.com/paritytech/polkadot-sdk/blob/436b4935b52562f79a83b6ecadeac7dcbc1c2367/substrate/client/transaction-pool/src/graph/validated_pool.rs#L53-L62).

---

> We just should increase metric with dedicated label (e.g. "validation_failure" + error).

I would go with this. What this entails is returning two separate sets from `revalidate_inner` in mempool (one with tx_hash/IntoPoolError and another one with tx_hash/common::error::Error). I'll see if there are other ways of returning "invalid" txs (tried to keep things simple without increasing the scope with too much refactoring), but for now I think what you mention makes sense.

So to summarise, your suggestion is to still keep the internal errors under `mempool_revalidation_invalid_txs` (which I am ok with). Separately, we'll log the internal error, and also treat the associated tx like in the past. Hope this is aligned with your understanding!